qiskit_optimization/applications/bin_packing.py

```diff
@@ -0,0 +1,149 @@
+
+try:
+    import matplotlib.pyplot as plt
+
+    _HAS_MATPLOTLIB = True
+except ImportError:
+    _HAS_MATPLOTLIB = False
+
+
+class BinPacking(OptimizationApplication):
+    """Optimization application for the "bin packing" [1] problem.
+
+    References:
+        [1]: "Bin packing",
+        `https://en.wikipedia.org/wiki/Bin_packing_problem
+        <https://en.wikipedia.org/wiki/Bin_packing_problem>`_
+    """
+
+    def __init__(
+        self, weights: List[int], max_weight: int, max_number_of_bins: Optional[int] = None
+    ) -> None:
+        """
+        Args:
+            weights: A list of the weights of items
+            max_weight: The maximum bin weight capacity
+            max_number_of_bins: The maximum number of bins by default equal to the number of items
+        """
+        self._weights = weights
+        self._max_weight = max_weight
+        if max_number_of_bins is None:
+            self._max_number_of_bins = len(weights)
+        else:
+            self._max_number_of_bins = max_number_of_bins
+
+    def to_quadratic_program(self) -> QuadraticProgram:
+        """Convert a bin packing problem instance into a
+        :class:`~qiskit_optimization.problems.QuadraticProgram`
+
+        Returns:
+            The :class:`~qiskit_optimization.problems.QuadraticProgram` created
+            from the bin packing problem instance.
+        """
+        mdl = Model(name="BinPacking")
+        num_bins = self._max_number_of_bins
+        num_items = len(self._weights)
+        y = mdl.binary_var_list(num_bins, name="y")
+        mdl.minimize(mdl.sum(y))
+        x = mdl.binary_var_matrix(num_items, num_bins, name="x")
+        for i in range(num_items):
+            # First set of constraints: the items must be in any bin
+            mdl.add_constraint(mdl.sum(x[i, j] for j in range(num_bins)) == 1)
+        for j in range(num_bins):
+            # Second set of constraints: weight constraints
+            mdl.add_constraint(
+                mdl.sum(self._weights[i] * x[i, j] for i in range(num_items))
+                <= self._max_weight * y[j]
+            )
+        op = from_docplex_mp(mdl)
+        return op
+
+    def interpret(self, result: Union[OptimizationResult, np.ndarray]) -> List[List[int]]:
+        """Interpret a result as item indices
+
+        Args:
+            result : The calculated result of the problem
+
+        Returns:
+            items_in_bins: A list of lists with the items in each bin
+        """
+        x = self._result_to_x(result)
+        num_items = len(self._weights)
+        num_bins = self._max_number_of_bins
+        bins = x[:num_bins]
+        items = np.array(x[num_bins:]).reshape((num_items, num_bins))
+        items_in_bins = [
+            [i for i in range(num_items) if bins[j] and items[i, j]] for j in range(num_bins)
+        ]
+        return items_in_bins
+
+    def draw(self, result: Union[OptimizationResult, np.ndarray]):
```

Comment:
This does not seem like the same as the `draw` methods in the graph applications that internally use nx.draw and actually draw the graph. This here seems more like a get_figure/get_plot type method since it returns something that you can then draw. 

Also as it stands it has no return typehint. Before it returned `plt.Figure` but I get that this does not work when MatPlotLib is not installed. Rather perhaps what can be done, assuming what is wanted here is to return a figure that the user can draw, rather than actually drawing it. is to do similar to what is done in TorchConnector in MachineLearing that references types that depend on Torch being installed and would otherwise fail if it is not. Basically this here could return Figure - that would be imported from matplotlib.plt but if that failed then it could just define a dummy Figure locally to satisfy the code. Since it will check whether matplotlib is installed this dummy class is simply to allow the code to get that far, it would not be used in practice.

Reply:
FYI: We use a fake class for a similar situation of Gurobi. https://github.com/Qiskit/qiskit-optimization/blob/d911c452b7e82bb9b48ed9655dc8fe9b01bfe6fa/qiskit_optimization/translators/gurobipy.py#L25

Reply:
Thank you @woodsp-ibm, I followed your suggestions.